### PR TITLE
lightning/br: move some common code into outer pkg

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #52182

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/backup/client.go [hunk 1]
```diff
@@ -443,55 +442,6 @@ func CheckBackupStorageIsLocked(ctx context.Context, s storage.ExternalStorage)
 	return nil
 }
 
-// BuildTableRanges returns the key ranges encompassing the entire table,
-// and its partitions if exists.
-func BuildTableRanges(tbl *model.TableInfo) ([]kv.KeyRange, error) {
```
Comment (D3Hunter) [inline br/pkg/backup/client.go:448]: moved into `distsql/request_builder.go`

File: br/pkg/redact/redact.go
Comment (D3Hunter) [inline br/pkg/redact/redact.go:1]: moved into `pkg/util/redact`

File: br/pkg/utils/db.go [hunk 2]
```diff
@@ -24,31 +23,9 @@ const (
...
 var (
-	// check sql.DB and sql.Conn implement QueryExecutor and DBExecutor
-	_ DBExecutor = &sql.DB{}
-	_ DBExecutor = &sql.Conn{}
-
 	logBackupTaskCount = atomic.NewInt32(0)
 )
 
-// QueryExecutor is a interface for exec query
-type QueryExecutor interface {
```
Comment (D3Hunter) [inline br/pkg/utils/db.go:35]: it's duplicate with code in `pkg/util/dbutil`, so remove this one

File: br/pkg/utils/env.go [hunk 3]
```diff
@@ -1,33 +0,0 @@
...
-package utils
-
-import (
-	"github.com/pingcap/log"
-	"go.uber.org/zap"
-	"golang.org/x/net/http/httpproxy"
-)
-
-// LogEnvVariables logs related environment variables.
-func LogEnvVariables() {
```
Comment (D3Hunter) [inline br/pkg/utils/env.go:12]: moved into pkg/util/logutil

File: br/pkg/utils/math.go [hunk 4]
```diff
@@ -1,16 +0,0 @@
-// Copyright 2020 PingCAP, Inc. Licensed under Apache-2.0.
-
-package utils
-
-// NextPowerOfTwo returns the smallest power of two greater than or equal to `i`
-// Caller should guarantee that i > 0 and the return value is not overflow.
-func NextPowerOfTwo(i int64) int64 {
```
Comment (D3Hunter) [inline br/pkg/utils/math.go:7]: moved into pkg/util/mathutil

File: br/pkg/lightning/common/util.go [hunk 5]
```diff
@@ -696,3 +697,19 @@ func IsRaftKV2(ctx context.Context, db *sql.DB) (bool, error) {
...
+
+	passwordRegexp *regexp.Regexp
+)
+
+func init() {
+	passwordRegexp = regexp.MustCompile(passwordPatterns)
+}
+
+// HideSensitive replace password with ******.
+func HideSensitive(input string) string {
```
Comment (D3Hunter) [inline br/pkg/lightning/common/util.go:712]: moved from br/pkg/utils/sensitive.go, we are the only one uses it

File: br/pkg/utils/worker.go [hunk 6]
```diff
@@ -8,109 +8,8 @@ import (
 	berrors "github.com/pingcap/tidb/br/pkg/errors"
 	"github.com/pingcap/tidb/br/pkg/logutil"
 	"go.uber.org/zap"
-	"golang.org/x/sync/errgroup"
 )
 
-// WorkerPool contains a pool of workers.
-type WorkerPool struct {
```
Comment (D3Hunter) [inline br/pkg/utils/worker.go:15]: moved into pkg/util/worker_pool.go

File: pkg/util/dbutil/dbutiltest/utils.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/pingcap/tidb/pkg/ddl"
+	"github.com/pingcap/tidb/pkg/parser"
+	"github.com/pingcap/tidb/pkg/parser/ast"
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/types"
+	"github.com/pingcap/tidb/pkg/util/dbutil"
+)
+
+// GetTableInfo returns table information.
+func GetTableInfo(ctx context.Context, db dbutil.QueryExecutor, schemaName string, tableName string) (*model.TableInfo, error) {
```
Comment (D3Hunter) [inline pkg/util/dbutil/dbutiltest/utils.go:32]: moved from dbutil to avoid import cycle

File: pkg/util/dbutil/table.go [hunk 8]
```diff
@@ -99,48 +38,3 @@ func FindColumnByName(cols []*model.ColumnInfo, name string) *model.ColumnInfo {
 
 	return nil
 }
-
-// EqualTableInfo returns true if this two table info have same columns and indices
-func EqualTableInfo(tableInfo1, tableInfo2 *model.TableInfo) (bool, string) {
```
Comment (D3Hunter) [inline pkg/util/dbutil/table.go:104]: only used in test, moved into test code

File: pkg/util/redact/redact.go [hunk 9]
```diff
@@ -172,3 +173,34 @@ func DeRedact(remove bool, input io.Reader, output io.Writer, sep string) error
...
+}
+
+// NeedRedact returns whether to redact log
+func NeedRedact() bool {
+	mode := errors.RedactLogEnabled.Load()
+	return mode != errors.RedactLogDisable && mode != ""
+}
+
+// Value receives string argument and return omitted information if redact log enabled
+func Value(arg string) string {
```
Comment (lance6716) [inline pkg/util/redact/redact.go:193]: I prefer this function to be renamed to `String` to indicate its argument is string type, more like zap.String.

File: pkg/util/redact/redact.go [hunk 10]
```diff
@@ -172,3 +173,34 @@ func DeRedact(remove bool, input io.Reader, output io.Writer, sep string) error
...
+	if redactLog {
+		mode = errors.RedactLogEnable
+	}
+	errors.RedactLogEnabled.Store(mode)
+}
+
+// NeedRedact returns whether to redact log
+func NeedRedact() bool {
+	mode := errors.RedactLogEnabled.Load()
+	return mode != errors.RedactLogDisable && mode != ""
```
Comment (lance6716) [inline pkg/util/redact/redact.go:189]: can we check it's RedactLogEnable?

File: pkg/util/redact/redact.go (hunk 9 above)
Comment (D3Hunter) [reply pkg/util/redact/redact.go:193]: this package already have a `String` with quite different impl

this method is actually redacting values inside SQL, i.e. convert values into `?`, so name it as `Value`

File: pkg/util/redact/redact.go (hunk 10 above)
Comment (D3Hunter) [reply pkg/util/redact/redact.go:189]: just move code, there's other values
```
	RedactLogEnable  string = "ON"
	RedactLogDisable        = "OFF"
	RedactLogMarker         = "MARKER"
```

**Comment by D3Hunter** [comment]
it's insane that we have a `init` set `newCollationEnabled=1`
https://github.com/pingcap/tidb/blob/732fee0c61c6efed9ec4df9398e892b60eeca582/pkg/util/collate/collate.go#L397-L398

and in here, we set to `newCollationEnabled=0` in `init` too

https://github.com/pingcap/tidb/blob/732fee0c61c6efed9ec4df9398e892b60eeca582/pkg/util/dbutil/table.go#L33-L35

File: pkg/util/dbutil/table.go [hunk 11]
```diff
@@ -15,78 +15,12 @@
...
 	"github.com/pingcap/tidb/pkg/parser/model"
-	_ "github.com/pingcap/tidb/pkg/planner/core" // to setup expression.EvalAstExpr. See: https://github.com/pingcap/tidb/blob/a94cff903cd1e7f3b050db782da84273ef5592f4/planner/core/optimizer.go#L202
-	"github.com/pingcap/tidb/pkg/types"
 	_ "github.com/pingcap/tidb/pkg/types/parser_driver" // for parser driver
-	"github.com/pingcap/tidb/pkg/util/collate"
 )
 
-func init() {
-	collate.SetNewCollationEnabledForTest(false)
-}
```
Comment (D3Hunter) [inline pkg/util/dbutil/table.go:35]: see https://github.com/pingcap/tidb/pull/52191#issuecomment-2024836481

after we reuse interface in this pkg in lightning, this `init` might cause some test fail

File: br/pkg/lightning/common/util.go [hunk 12]
```diff
@@ -696,3 +697,19 @@ func IsRaftKV2(ctx context.Context, db *sql.DB) (bool, error) {
 	}
 	return false, rows.Err()
 }
+
+var (
+	passwordPatterns = `(password[\s]*=[\s]*(\\")?)(.*?)((\\")?\\n)`
+
+	passwordRegexp *regexp.Regexp
```
Comment (lance6716) [inline br/pkg/lightning/common/util.go:704]:
```suggestion
	passwordRegexp  = regexp.MustCompile(passwordPatterns)
```

**Review by lance6716 — APPROVED** [review]

**Review by lyzx2001 — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Review by 3pointer — APPROVED** [review]

**Pull request merged**